index.ts

```diff
@@ -1,213 +1,170 @@
+      const userId = authEvent.pubkey
+      const userRole = env.BANBOORU_PUBKEY_ROLE_KV.get(authEvent.pubkey) as Promise<Role | undefined>
+      return next(request, { ...env, authEvent, userId, userRole }, params)
     }
-
-    const path_file_sha256_hex = url.pathname.slice(6);
-    if (!path_file_sha256_hex.match(/^[A-Fa-f0-9]{64}$/)) {
-      return new Response("/file/<FILE-SHA-256> could not validate SHA-256", {
-        status: 400,
-      }); // 400 Bad Request
+  }
+}
+
+/**
+ * Restricted access middleware. Requires a NIP-94 Authorization header from a known pubblic key with one of the specific `roles`.
+ */
+function restricted(roles: Role[], next: MatchHandler<AuthEnv>): MatchHandler<WorkerEnv> {
+  return auth(async (request: Request, env: HandlerContext<AuthEnv>, params: Record<string, string>) => {
+    const role = await env.userRole
+    if(role && roles.includes(role)) {
+      return next(request, env, params)
+    } else {
+      // TODO: Return JSON body with error message.
+      return new Response("Forbidden.", { status: 403 });
     }
+  })
+}
 
-    // GET /file/<SHA-256>
-    if (request.method === "GET") {
-      const object = await env.BANBOORU_BUCKET.get(path_file_sha256_hex);
+export default {
+  fetch: router<WorkerEnv>({
+    'GET /file/:hash': async function getFile(_req, env, params) {
+      if(!params.hash.match(hexKeyPattern)) return new Response(`Invalid SHA256 hash: ${params.hash}`, { status: 400 });
+      const object = await env.BANBOORU_BUCKET.get(params.hash);
       if (object === null) {
         return new Response("Object Not Found", { status: 404 });
       }
       const headers = new Headers();
       object.writeHttpMetadata(headers);
       headers.set("etag", object.httpEtag);
       return new Response(object.body, { headers });
-    }
-
-    // PUT /file/<SHA-256>
-    if (request.method === "PUT") {
-      const auth_event: Event = await get_auth_event(request);
-      if (!auth_event) {
-        return new Response("Unauthorized.", { status: 401 }); // 401 Unauthorized
+    },
+    'PUT /file/:hash': restricted(["admin", "user"], async function putFile(request, env, params) {
+      if(!params.hash.match(hexKeyPattern)) return new Response(`Invalid SHA256 hash: ${params.hash}`, { status: 400 });
+      if(!request.body) return new Response("Missing body", { status: 400 });
+  
+      // We expect the SHA256 hash from the URL to match the one of the request body (ie. the uploaded file).
+      // Per NIP-98 the authentication SHOULD include a SHA256 hash of the request body in a payload tag as hex (["payload", "<sha256-hex>"]).
+      // We will validate that the actual payload hash match when uploading to R2.
+      const payloadTag = env.authEvent.tags.find((t: string[]) => t[0] === 'payload')
+      if(payloadTag && payloadTag[1] && payloadTag[1] != params.hash) {
+        return new Response(`Unauthorized: Invalid nostr event, payload signature invalid`, { status: 401 });
```

Comment:
This might provide a little more clarity:
```
return new Response(`Unauthorized: Invalid nostr event, payload hash does not match params hash`, { status: 401 });
```